langchain4j-elasticsearch/src/main/java/dev/langchain4j/store/embedding/elasticsearch/ElasticsearchEmbeddingStoreImpl.java

```diff
@@ -0,0 +1,226 @@
+                    SearchRequest.of(s -> s.query(n -> n.scriptScore(scriptScoreQuery)).size(maxResults)), Document.class);
+
+            return response.hits().hits().stream()
+                    .map(hit -> Optional.ofNullable(hit.source()).map(document -> new EmbeddingMatch<>(hit.score(), hit.id(),
+                                    new Embedding(document.getVector()),
+                                    new TextSegment(document.getText(), Optional.ofNullable(document.getMetadata()).map(Metadata::new).orElse(null))))
+                            .orElse(null)).collect(Collectors.toList());
+        } catch (IOException e) {
+            log.error("[ElasticSearch encounter I/O Exception]", e);
+        } catch (ElasticsearchException e) {
+            log.error("[ElasticSearch Exception]", e);
+        }
+
+        return new ArrayList<>();
+    }
+
+    private void addInternal(String id, Embedding embedding, TextSegment embedded) {
+        addAllInternal(Collections.singletonList(id), Collections.singletonList(embedding), embedded == null ? null : Collections.singletonList(embedded));
+    }
+
+    private void addAllInternal(List<String> ids, List<Embedding> embeddings, List<TextSegment> embedded) {
+        if (isCollectionEmpty(ids) || isCollectionEmpty(embeddings)) {
+            log.info("[do not add empty embeddings to elasticsearch]");
+            return;
+        }
+        ValidationUtils.ensureTrue(ids.size() == embeddings.size(), "ids size is not equal to embeddings size");
+        ValidationUtils.ensureTrue(embedded == null || embeddings.size() == embedded.size(), "embeddings size is not equal to embedded size");
+
+        try {
+            createIndexIfNotExist(embeddings.get(0).length());
+
+            bulk(ids, embeddings, embedded);
+        } catch (IOException e) {
+            log.error("[ElasticSearch encounter I/O Exception]", e);
+        } catch (ElasticsearchException e) {
+            log.error("[ElasticSearch Exception]", e);
+        }
+    }
+
+    private void createIndexIfNotExist(int dim) throws IOException {
+        try {
+            client.indices().get(c -> c.index(indexName));
+        } catch (ElasticsearchException e) {
+            if (String.format("no such index [%s]", indexName).equals(e.response().error().reason())) {
+                client.indices().create(c -> c.index(indexName)
+                        .mappings(getDefaultMappings(dim)));
+            } else {
+                log.error("[Encounter unexpect exception when check index exist]", e);
+            }
+        }
+    }
+
+    private TypeMapping getDefaultMappings(int dim) {
+        // do this like LangChain do
+        Map<String, Property> properties = new HashMap<>(4);
+        properties.put("text", Property.of(p -> p.text(TextProperty.of(t -> t))));
+        properties.put("vector", Property.of(p -> p.denseVector(DenseVectorProperty.of(d -> d.dims(dim)))));
```

Comment:
In my opinion, `TypeMapping` is only necessary for fields that require strong type assurance. for `vector` we should ensure that it's a `DenceVector` so that we could use cosine similarity to calculate score.

for metadata, actual it's a `JSON` type and elasticsearch can auto detect it's mapping, so we don't have to specify it manually. (But you can specify if you want)

Reply:
got it, thanks for explanation!